core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -577,9 +577,12 @@ private RexNode simplifyIsNull(RexNode a) {
   private RexNode simplifyCoalesce(RexCall call) {
     final Set<String> digests = new HashSet<>();
     final List<RexNode> operands = new ArrayList<>();
+    final RexSimplify simplify = withUnknownAsFalse(false);
     for (RexNode operand : call.getOperands()) {
-      operand = simplify_(operand);
-      if (digests.add(operand.digest)) {
+      operand = simplify.simplify_(operand);
+      if (digests.add(operand.digest)
+          && (!RexUtil.isNull(operand)
+                  || !unknownAsFalse)) {
```

Comment:
Are those `!RexUtil.isNull(operand)|| !unknownAsFalse` checks really required?

Reply:
In unknownAsFalse mode,without this check,the "Coalesce(null,x)" will not simplify to "x".I can  
 change this to "!(RexUtil.isNull(operand)&& unknownAsFalse)" which is more understandable.

Reply:
This is an optimization feature that is not listed in CALCITE-2580, neither it is covered in the tests.

I suggest to trim "leading nulls" to a separate JIRA.

Reply:
Ok